### PR TITLE
Revamp Tiberian Dawn ingame UI

### LINKED ISSUE #7572: TD UI
Just some ideas, mock-ups that I have created for a possible TD UI.

chrometd2
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/9088015-td-ui?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by dragunoff**
This PR implements a revamped ingame UI for Tiberian Dawn. The UI is based on work done by @xan2622 and @mabulsoud with some twists from myself. The observer UI remains the same, that could be tackled in a future PR.

Closes #7572 
PR with design source files: OpenRA/ArtSrc#7

## Changes
The second commit is quite large so here's a summary of the changes:

**Images:**
* Revamp `chrome.png`
* ~~Add `dialog.png` (split from `chrome.png`)~~
* ~~Add `bits.png` (split from `chrome.png`)~~
* Remove `strategic.png` (moved into `bits.png`)

**YAMLs:**
* Revamp `chrome.yaml` and `ingame.yaml`
* Minor changes to `mod.yaml`, `editor.yaml` and `ingame-chat.yaml`

**cs:**
* Upadte image coordinates in `CncLoadScreen.cs`

## GDI
gdi-chrome
* GDI tooltips remain unchanged due to some limitations of the widgets code

## Nod
nod-chrome

## Links
[PSD source files](https://drive.google.com/drive/folders/1AiECqjByN9XZ_XpRpByo-xIHZYy2SdQG?usp=sharing)
[Forum thread](https://forum.openra.net/viewtopic.php?f=82&t=20849)

### CODE REVIEW

**Comment by pchote** [comment]
I raised this issue in the earlier discussion, and I'm afraid I'm going to have to insist on it now that it has made its way to a PR: chrome.png is generated from [chrome.svg](https://github.com/OpenRA/ArtSrc/blob/master/cnc/chrome.svg) and should never be edited by hand.

Please update chrome.svg to add anything that can be vector based, and then create a new 512x512 raster format file (ideally xcf or psd without any photoshop-specific features) for the pixel based sidebar backgrounds.

**Comment by dragunoff** [comment]
Thanks @pchote ☺️ I will wait for any usability reviews before addressing the technical matters. I was not aware that chrome.svg existed. I'm not sure that I'd be able to pull it off entirely without Photoshop effects but I'll see what can be done. 

The duplication of assets was intentional because I based that on TS chrome and I wanted to have distinct skins for each faction. That idea didn't quite work out but I decided to keep the layout.

Is it OK that I split off `bits.png` and `dialog.png`? I assume that I'll need to file a PR to the ArtSrc repo with source files for them?

**Comment by pchote** [comment]
Hmm, I was sure that we had already talked about chrome.svg, but I must have been mistaken.

> I'm not sure that I'd be able to pull it off entirely without Photoshop effects but I'll see what can be done.

The new elements are fine as raster art if that is the primary source.   My main annoyance was with the glyphs and other pre-existing artwork.

> Is it OK that I split off bits.png and dialog.png?

Texture samplers are even more precious than VRAM - rendering performance falls off a cliff if we use more than 8 textures in a frame, and this includes the texture atlases used for rendering everything in the world. Textures should only be split off if the benefits outweigh a potentially non-neglible performance degradation.

**Comment by Inq8** [comment]
55062705-31ea0600-507f-11e9-864e-33e27633a73b

Great work, ultimate nitpicking but I don't think that highlighted icon is centered correctly, or the button itself is too big. (visually its sitting off to the left)

I think the defence stance is also not centered, although it seems less obvious.

**Comment by dragunoff** [comment]
> Great work, ultimate nitpicking but I don't think that highlighted icon is centered correctly, or the button itself is too big. (visually its sitting off to the left)
> 
> I think the defence stance is also not centered, although it seems less obvious.

You are correct! Good catch! The buttons have uneven width. I will fix that.

**Comment by pchote** [comment]
Some UI comments, mainly focused on the command bar:
* IMO the frame around the command bar is too chunky. It looks much nicer if the frame width is reduced by 2px:
   reduced
 * The stance buttons feel IMO too narrow, making them harder to select with the mouse. Can we adopt the same button sizes used in RA?
 * The chat dialog needs to move up a bit so that it doesn't overlap with the commandbar at narrower window sizes.
 * We still lose at Fitt's law by having the bar offset from the edge and having an unclickable frame - but this applies to all mods so IMO is out of scope of this PR.
* C&Cs have always had the powerbar on the left, so having it swapped with the silo bar here feels wrong. Please swap them back.
* The red button separator lines on the Nod sidebar feel too strong. Can we tone these down somehow?
* IMO the faction logos feel a bit small in the radar bin. Is there a reason not to make them bigger?

**Comment by dragunoff** [comment]
Thanks for the review @pchote -- I have already started working on the technical side (reducing the number of sprites and the overall size of the sprites). I mostly agree with your comments and I will look into fixing the issues. 
> 
>     * IMO the frame around the command bar is too chunky. It looks much nicer if the frame width is reduced by 2px:
> 
>     * The stance buttons feel IMO too narrow, making them harder to select with the mouse. Can we adopt the same button sizes used in RA?
> 
>     * The chat dialog needs to move up a bit so that it doesn't overlap with the commandbar at narrower window sizes.
> 
>     * We still lose at Fitt's law by having the bar offset from the edge and having an unclickable frame - but this applies to all mods so IMO is out of scope of this PR.
> 

I agree with everything here. 

>     * C&Cs have always had the powerbar on the left, so having it swapped with the silo bar here feels wrong. Please swap them back.
> 

I was aiming for consistency with RA and the cash count seemed better with "$" at the front. However you're right that it feels wrong for CNC.

>     * The red button separator lines on the Nod sidebar feel too strong. Can we tone these down somehow?
> 

Sure thing. I will try the darker red used on the UI panel borders.

>     * IMO the faction logos feel a bit small in the radar bin. Is there a reason not to make them bigger?
> 

They look quite fine to me but I'll try making them a little larger.

**Comment by pchote** [comment]
Don't feel the need to go overboard with micro-optimizing the sprite layout, just not being overly wasteful with duplication/empty space/new files should be good enough here.

**Comment by dragunoff** [comment]
I've already changed the old sprite a quite a lot. I want to make it more logical with separate areas for icons, dialog boxes, logos, etc. I will file a PR to the ArtSrc repo soon.

**Comment by pchote** [comment]
Any new progress here @dragunoff? We're not far off a new playtest cycle, and it would be a shame to miss out on this.

**Comment by dragunoff** [comment]
@pchote Yes, I'm working on it :-) I have combined everything into a single 1024x512 sprite and I'm working through the changes requested. Expect an update until the end of this week.

**Comment by dragunoff** [comment]
I have addressed all the comments and updated this PR and filed OpenRA/ArtSrc#7 with the design source files. Here are some new screenshots:

cnc-ui-screenshots-gdi

cnc-ui-screenshots-nod

**Comment by dragunoff** [comment]
Updated because the performance graph was overlaping the command bar.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/AddFactionSuffixLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/AddFactionSuffixLogic.cs]: This can simplify to `if (widget is ButtonWidget)` and likewise for the other checks below.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/AddFactionSuffixLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/AddFactionSuffixLogic.cs]: These lines can then simplify to a direct cast and set: `((ButtonWidget)widget).Background += suffix;`

**Comment by pchote** [comment]
> GDI tooltips remain unchanged due to some limitations of the widgets code

Fixed with 94793ca85b3a33b81c4f62a52c57c57fcda2be9c - please cherry pick this into the PR.

The ingame chat will need some deeper changes to the `TextFieldWidget` so is best left to a followup PR.

Can you please also rebase this on latest bleed? The GitHub UI isn't reporting a merge conflict, but this will break things if merged from the current base. Beware of #16530 though, which will affect this.

**Comment by pchote** [comment]
I left a couple of requests on https://github.com/OpenRA/ArtSrc/pull/7.

This is looking really great, and none of the requests should hopefully be too difficult, so I think this should be good to merge once they're done.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/AddFactionSuffixLogic.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/AddFactionSuffixLogic.cs]: Fixed.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/AddFactionSuffixLogic.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/AddFactionSuffixLogic.cs]: Fixed.

**Comment by dragunoff** [comment]
Rebased, cherry-picked and updated the sprite. Here's the a screenshot with faction tooltips in:

td-ui-revamp

**Comment by dragunoff** [comment]
Update to fix code style error.

**Comment by dragunoff** [comment]
Updated with the latest changes to `chrome.svg`.

**Review by pchote — APPROVED** [review]
LGTM :+1:

This suffers from #16530 but I would prefer we deal with that at the same time as the rest of the regressions instead of implementing something here that may need to be reverted or changed.

If I wanted to nit-pick I could raise some points about the chrome definition naming and unused faction variants, but this would open up a much larger can of worms that I don't think is worth dealing with.

**Comment by dragunoff** [comment]
> If I wanted to nit-pick I could raise some points about the chrome definition naming and unused faction variants, but this would open up a much larger can of worms that I don't think is worth dealing with.
>
I deliberately included faction variants for all panels even though they are not used. I thought I'd better do it now that I am onto it. It may come in handy if we want to do anything with faction suffix later.

I noticed a few other unused definitions in `chrome.yaml` but I was not sure what the intentions were behind them so I left them untouched.

**Comment by pchote** [comment]
a much larger can of worms :smile:

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0a2a738ba71d627ace98e26d64dd5185a380ddf1...ef571ad8aadba2e206c038e416b37573012dbc60

**Pull request merged**
